src/irmin/irmin.ml

```diff
@@ -181,28 +158,37 @@ struct
     module Repo = struct
       type t = {
         config : Conf.t;
-        values : [ `Read ] Values.t;
+        contents : [ `Read ] Contents.t;
+        nodes : [ `Read ] Node.t;
+        commits : [ `Read ] Commit.t;
         branch : Branch.t
       }
 
-      let contents_t t = t.values
+      let contents_t t = t.contents
 
-      let node_t t = (t.values, t.values)
+      let node_t t = t.nodes
 
-      let commit_t t = ((t.values, t.values), t.values)
+      let commit_t t = t.commits
 
       let branch_t t = t.branch
 
       let batch t f =
-        Values.batch t.values @@ fun t ->
-        let contents_t = t in
-        let node_t = (contents_t, t) in
-        let commit_t = (node_t, t) in
+        Contents.CA.batch t.contents @@ fun c ->
+        Node.CA.batch (snd t.nodes) @@ fun n ->
+        Commit.CA.batch (snd t.commits) @@ fun ct ->
```

Comment:
The API is not ideal here, as we might flush 3 times -- might be avoided if the backend is "clever". The previous code (where we add only one store) allowed to avoid this by considering everything as string. But it was buggy because the `pre_digest` custom functions were not called.